### PR TITLE
Indexing of TMA

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR adds the indexing of TMA, please start the review from note `[Schedule and indexing of TMA]` to get an idea on how the schedule looks like and what is the algorithm for indexing.

With this PR, we are no longer limited to loading the entire tensor as a tile. We can arbitrarily view the tensor and create tiles, and arbitrarily schedule the tile and non-tile IterDomains.

Other limitations still apply, for example, we are still limited to only one thread per block. And to make this PR small and incremental, I decided to leave some important tasks for the future:
- Validation, test of invalid cases
- Investigation of some bugs

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3368]: This is indeed `(6, 6)`. I think it is mostly a matter of the definition of "tile size". Here, tile size 8 does not mean you have in total 8 items, but instead, you have a span of 8, and you select odd numbers that is >=0 and < 8.

Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3368]: Is "tile size" confusing? Should we think of another name like "box dim" (just like what the driver API is using)?

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3356]: You are right. It's wrong to say `if it has no parent`.  This was from my previous design.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3604]: changed

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3355]: changed

File: tests/cpp/test_memory.cpp
Comment (zasdfgbnm) [inline tests/cpp/test_memory.cpp:266]: I can confirm that, if I add a
```C++
asm volatile("nanosleep.u32 10000;");
```
before the `Hopper::cpAsyncBulkTensorTileS2G`, the test will pass.

Comment (zasdfgbnm) [reply tests/cpp/test_memory.cpp:266]: Adding a `asm volatile("fence.proxy.async.shared::cta;");` also resolve the problem.

Comment (zasdfgbnm) [reply tests/cpp/test_memory.cpp:266]: fixed by https://github.com/NVIDIA/Fuser/pull/1967

**Comment by zasdfgbnm** [comment]
!build

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3367]: Sure, this is certainly a better description. What do you mean by
> for two outer-most ranks in the structure.

isn't there 2 ranks in total, so it should be like below?

> for two ranks in the structure.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3503]: Good question. I have no idea how C++'s `[=]` works with reference. Looks like it automatically dereference:

```C++
#include <iostream>
#include <functional>
#include <vector>

class A {
public:
    A() {
        std::cout << "ctor " << this << std::endl;
    }
    A(const A& a) {
        std::cout << "copy ctor " << &a << " -> " << this << std::endl;
    }
    ~A() {
        std::cout << "dtor " << this << std::endl;
    }
    void print() const {
        std::cout << "A::print " << this << std::endl;
    }
};

int main() {
    std::function<void()> f;
    std::cout << "enter scope" << std::endl;
    {
        std::vector<A> av(1);
        const A& a = av[0];
        f = [=]() {
            a.print();
        };
    }
    std::cout << "leave scope" << std::endl;
    f();
}
```

```
enter scope
ctor 0x986ec0
copy ctor 0x986ec0 -> 0x7ffc00b40667
copy ctor 0x7ffc00b40667 -> 0x986ee0
dtor 0x7ffc00b40667
dtor 0x986ec0
leave scope
A::print 0x986ee0
dtor 0x986ee0
```

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3379]: I added a few paragraphs before `Definition 1` explaining how the schedule looks like https://github.com/NVIDIA/Fuser/pull/1962/commits/7718921da6b430dfd94d2b0003ed35586388290c. Hope that will help answer you question. For your question, if the TensorView is 1D, but you want to use a 2D TMA, then you will need to first view it as 2D then create tile. So you will have something like:

```
  I1 -> split -> I2 -> split -> I4 (Bulk)
            |               \-> I5 (non-Bulk)
             \-> I3 -> split -> I6 (Bulk)
                            \-> I7 (non-Bulk)
```

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3597]: > inputs will be bulk due to the fact that we are replaying transformations from consumer to producer? 

No, this has nothing to do with replaying producer as consumer. We replay producer as consumer because the consumer dictate the loop structure of an expression. For example, if we have `T1[I1*I2] = T0[I1, I2]`, we will have loop structure

```python
for i in I1*I2:
  T1[i] = T0[i / size2, i % size2]
```

instead of

```python
for i1 in I1:
  for i2 in I2:
    T1[i1 * size2 + i2] = T0[i1, i2]
```

Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3597]: I think the confusion here is the concept "bulk IterDomain" vs "IterDomain parallelized as `ParallelType::Bulk`", they are different concepts. An "IterDomain parallelized as `ParallelType::Bulk`" is always a "bulk IterDomain", but not the reverse. We only parallelize leaf domains, so in your first diagram, it is an error if you parallelize `I2` as `ParallelType::Bulk`. It is a "bulk IterDomain" but not an "IterDomain parallelized as `ParallelType::Bulk`".

Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3603]: I added some texts explaining these two concepts: https://github.com/NVIDIA/Fuser/pull/1962/commits/bfa3ddbcca8d24587b38b3427d0a1076c35d256a

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3649]: What if we have an invalid fusion like this?
```
// [I1, I2]
tv->swizzle(XOR, 0, 1);
// [I1', I2']
tv->axis(1)->parallelize(ParallelType::Bulk);
// [I1', I2' (Bulk) ]
```
Then `I2'` is an "originating bulk IterDomain", but this schedule is invalid.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3756]: > with multiple `split` Exprs the stride of outer most entry in `frontier` may be several split factors larger than allocation domain stride.

What is the problem of this? For example, if we have an 1D tensor with shape `(102400,)`, and stride `(1,)`, and we split the tensor as `(100, 1024)`, shouldn't have stride `(1024, 1)`?

File: csrc/index_compute.cpp
Comment (naoyam) [inline csrc/index_compute.cpp:3308]: nit: `<= 5D`

File: csrc/index_compute.cpp
Comment (naoyam) [inline csrc/index_compute.cpp:3336]: Just a quick question. This part appears to indicate `I4` and `I7` are used together to generate the five "Bulk" domains. Is that correct? Or, are `I4` and `I7` transformed independently?

Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3336]: They can be independent or together, it makes no difference.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: fixed

File: csrc/index_compute.cpp
Comment (drzejan2) [reply csrc/index_compute.cpp:3530]: Thanks for checking this! I should have done it by myself, sorry for that.

It's good to know what is the behaviour, especially in cases where copy ctor might be expensive.

File: csrc/index_compute.cpp
Comment (drzejan2) [reply csrc/index_compute.cpp:3385]: Thanks for the example of scheduling 1D TV into 2d tiled bulk memory copy one. Now, that I see it, it obvious that it should be done this way.

The changes in the description look also good. Thank you!

File: csrc/index_compute.cpp
Comment (drzejan2) [reply csrc/index_compute.cpp]: > What do you mean by
> > for two outer-most ranks in the structure.
>
> isn't there 2 ranks in total, so it should be like below?
> > for two ranks in the structure.

You are right, my bad, I wrote this part of comment at early stage of review, before I reread the documentation.


To be double check, if I understood everything correctly, based on the description you have added, the relationship between IterDomains looks like this:
```
I1 -> split -> I2 (non-bulk) -> split -> I4 (non-bulk) -> split -> I8 (non-bulk)
          |                         |                          \-> I9 (bulk)
          |                         \--> I5 (non-bulk)
          |
          \--> I3 (non-bulk) -> split -> I6 (non-bulk) -> split -> I10 (non-bulk)
                                    |                          \-> I11 (bulk)
                                    \--> I7 (non-bulk)
```

Here the originating bulk IterDomains: `I9` and `I11`
  - both originating bulk IterDomains are outer products of split
  - their sibling Ids are non-bulk
Parents of originating IterDomains: `I4` and `I6`
  - both parents are iner products of split
  - parents' sibling IDs are non-bulk

Tile size: extends of `I9` and `I11`
Box size: extends of `I4` and `I6`


If that's correct, then this thread can be resolved, thanks!

File: csrc/index_compute.cpp
Comment (drzejan2) [inline csrc/index_compute.cpp:3444]: A minor suggestion, given the number of new terms introduced in this comment, I would find it really useful to see a set of examples with asii-art diagrams followed by explanation which IterDomains are originating bulk, box, global - and what will be passed to `cuTensorMapEncodeTiled`.

We could have examples for:
- 1D tensor - tiled to 2d and scheduled as 2d dense TMA transaction (box == tile)
- 1D tensor - tiled to 2d and scheduled as 2d sparse TMA transaction (box != tile)
- 2D tensor - tiled, but with merge expres included, scheduled as 2d dense TMA transaction

@zasdfgbnm , what do you think about this idea? Something similar to what I included in my [comment](https://github.com/NVIDIA/Fuser/pull/1962#discussion_r1533581104)

File: csrc/index_compute.cpp
Comment (drzejan2) [reply csrc/index_compute.cpp:3603]: Thank you @zasdfgbnm for explanation, especially for adding explanation about the difference between "bulk IterDomain" and "IterDomain parallelized as Bulk".

In the first diagram, I though this operation is legal, and the aim of such operation was to set parallelization type once, and have it automatically propagate to ancestor IterDomains, created when we apply merge/split.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: Yes, your understanding is correct.

File: csrc/index_compute.cpp
Comment (drzejan2) [reply csrc/index_compute.cpp:3649]: In this invalid fusion, won't the definition of `I2'` be `Swizzle2D` expression, and due to this definition, `I2'` will be rejected when we populate `originating_bulk_ids`, several lines above?

This is not a blocker, I was surprised see the same / similar check - feel free to resolve this thread :)

File: csrc/index_compute.cpp
Comment (drzejan2) [reply csrc/index_compute.cpp:3756]: Ok, now I understand my mistake. The inner most dimension, if tensor is dense, will have stride of `1` - for some reason I assumed it is the extend of the inner most dimension.

A fundamental mistake - sorry for the noise!

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3649]: Oh, I see what you mean. Yes, you are right. This check is redundant.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3444]: Good suggestion! I will add examples.

File: tests/cpp/test_memory.cpp
Comment (zasdfgbnm) [inline tests/cpp/test_memory.cpp:259]: I guess this should be already working. But I still want to leave this for the future to check. I am trying to keep the mental effort for this PR limited to the design and implementation of the schedule.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: removed

File: tests/cpp/test_memory.cpp
Comment (drzejan2) [inline tests/cpp/test_memory.cpp:557]: Just a though:

In the test it looks as if IterDomains' contiguity doesn't play any role, but I'm not sure how we could check if finalized TMA structures correctly handled strides and indices.

**Review by drzejan2 — APPROVED** [review]
I checked all changes, indexing seems to be reasonable, given my limited experience with that part of the system.

@zasdfgbnm Thank you for preparing this MR.

**Comment by zasdfgbnm** [comment]
I will wait for another approval from @naoyam. No hurry. I am not blocked on this.

**Comment by zasdfgbnm** [comment]
Converting this to draft to avoid accidental merge. But this is ready for review!

File: tests/cpp/test_memory.cpp
Comment (zasdfgbnm) [reply tests/cpp/test_memory.cpp:557]: If the result is correct, isn't it sufficient to convince us that the strides and indices are correctly handled?

File: csrc/index_compute.cpp
Comment (naoyam) [inline csrc/index_compute.cpp:3350]: Where's the definition of the box dim and the strided box dim?

Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3350]: It is in Definition 3: "boxId"

Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3350]: Updated the comment to refer to Definition 3

File: csrc/index_compute.cpp
Comment (naoyam) [inline csrc/index_compute.cpp:3321]: Should we define this part of the domains, i.e., `{I1, I2}`, as the TMA domain? Then, we could define the box domains as the inner split output domains of the TMA domain.

IIUC, the global domains defined below is the same as the TMA domain, right?

File: tests/cpp/test_memory.cpp
Comment (zasdfgbnm) [reply tests/cpp/test_memory.cpp]: I eventually decided to do it now. And as expected, it is working.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3321]: > IIUC, the global domains defined below is the same as the TMA domain, right?

Yes

> Should we define this part of the domains, i.e., `{I1, I2}`, as the TMA domain? Then, we could define the box domains as the inner split output domains of the TMA domain.

I do think we should give `{I1, I2}` a name, and TMA domain do sound like a good name. But conceptually, is it possible to **define** the box domain as a split of TMA domain? I think we can only **understand** the box domain as a split of TMA domain, but not **define** it this way. We need to define "global ID" first, because otherwise, how would you describe what IDs are included in TMA domain?

File: tests/cpp/test_memory.cpp
Comment (drzejan2) [reply tests/cpp/test_memory.cpp:556]: Ok, that's good enough proof that tested functionality works. I was just thinking about additional set of checks, but again, I don't have now a suggestion how it can be done.

I will mark this thread as resolved.

File: csrc/index_compute.cpp
Comment (naoyam) [inline csrc/index_compute.cpp:3464]: Could you give here an overview of the approach taken in this function? Would be helpful before diving into the function.

File: csrc/index_compute.cpp
Comment (naoyam) [reply csrc/index_compute.cpp:3321]: I just meant to say we name the split domain of a TMA domain as the box domain. Yes, we would first need to identify what domains should be included in the TMA domain.

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp]: Added in https://github.com/NVIDIA/Fuser/pull/1962/commits/62dd8b9fb72872b21336b4db128892c7fb6dae1e

File: csrc/index_compute.cpp
Comment (zasdfgbnm) [reply csrc/index_compute.cpp:3321]: updated in https://github.com/NVIDIA/Fuser/pull/1962/commits/62dd8b9fb72872b21336b4db128892c7fb6dae1e

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**
